aten/src/ATen/core/Array.h

```diff
@@ -16,10 +17,15 @@ struct alignas(16) Array {
   C10_HOST_DEVICE T& operator[](int i) {
     return data[i];
   }
-
-  C10_HIP_HOST_DEVICE Array() = default;
-  C10_HIP_HOST_DEVICE Array(const Array&) = default;
-  C10_HIP_HOST_DEVICE Array& operator=(const Array&) = default;
+#ifdef __HIP_PLATFORM_HCC__
```

Comment:
Was this ifdef necessary? `C10_HOST_DEVICE` is empty if `__HIP_PLATFORM_HCC__` isn't defined

Reply:
Yeah this was necessary when I moved it to ATen/core as is. Without this change, I was getting the following error:
```
23:21:52 [ 43%] Building CXX object caffe2/CMakeFiles/caffe2.dir/__/aten/src/ATen/CPUGenerator.cpp.o
23:21:52 [ 43%] Building CXX object caffe2/CMakeFiles/caffe2.dir/__/aten/src/ATen/CPUTypeDefault.cpp.o
23:21:52 [ 43%] Building CXX object caffe2/CMakeFiles/caffe2.dir/__/aten/src/ATen/Context.cpp.o
23:21:52 [ 43%] Building CXX object caffe2/CMakeFiles/caffe2.dir/__/aten/src/ATen/DLConvertor.cpp.o
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/CPUGenerator.cpp:1:
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/CPUGenerator.h:5:
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/PhiloxRNGEngine.h:15:
23:21:53 /var/lib/jenkins/workspace/aten/src/ATen/core/Array.h:21:3: error: unknown type name '__host__'
23:21:53   C10_HIP_HOST_DEVICE Array() = default;
23:21:53   ^
23:21:53 /var/lib/jenkins/workspace/c10/macros/Macros.h:189:29: note: expanded from macro 'C10_HIP_HOST_DEVICE'
23:21:53 #define C10_HIP_HOST_DEVICE __host__ __device__
23:21:53                             ^
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/CPUGenerator.cpp:1:
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/CPUGenerator.h:5:
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/PhiloxRNGEngine.h:15:
23:21:53 /var/lib/jenkins/workspace/aten/src/ATen/core/Array.h:21:22: error: expected ';' at end of declaration list
23:21:53   C10_HIP_HOST_DEVICE Array() = default;
23:21:53                      ^
23:21:53                      ;
23:21:53 /var/lib/jenkins/workspace/aten/src/ATen/core/Array.h:22:3: error: unknown type name '__host__'
23:21:53   C10_HIP_HOST_DEVICE Array(const Array&) = default;
23:21:53   ^
23:21:53 /var/lib/jenkins/workspace/c10/macros/Macros.h:189:29: note: expanded from macro 'C10_HIP_HOST_DEVICE'
23:21:53 #define C10_HIP_HOST_DEVICE __host__ __device__
23:21:53                             ^
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/CPUGenerator.cpp:1:
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/CPUGenerator.h:5:
23:21:53 In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/PhiloxRNGEngine.h:15:
23:21:53 /var/lib/jenkins/workspace/aten/src/ATen/core/Array.h:22:3: error: duplicate member '__device__'
23:21:53 /var/lib/jenkins/workspace/c10/macros/Macros.h:189:38: note: expanded from macro 'C10_HIP_HOST_DEVICE'
23:21:53 #define C10_HIP_HOST_DEVICE __host__ __device__
23:21:53                                      ^
23:21:53 /var/lib/jenkins/workspace/aten/src/ATen/core/Array.h:21:3: note: previous declaration is here
23:21:53   C10_HIP_HOST_DEVICE Array() = default;
23:21:53   ^
23:21:53 /var/lib/jenkins/workspace/c10/macros/Macros.h:189:38: note: expanded from macro 'C10_HIP_HOST_DEVICE'
23:21:53 #define C10_HIP_HOST_DEVICE __host__ __device__
```

Reply:
This is in the ROCm build? This doesn't make any sense. The ifdef around here exactly matches the ifdef in Macros.h. I'm concerned that something is messed up in the includes or macros.

https://github.com/pytorch/pytorch/blob/556af7c19db90465703a9acc0b437cbcfb018ee0/c10/macros/Macros.h#L188-L192

Reply:
May be the ifdef in the Macros should be __HIPCC__